db/warehouse/migrate/20230428203806_add_external_id_namespace.rb

```diff
@@ -0,0 +1,14 @@
+class AddExternalIdNamespace < ActiveRecord::Migration[6.1]
+  # ensure all IDs are unique in a namespace.
+  # https://www.pivotaltracker.com/story/show/185000178/comments/236605163
+  def change
+    add_column :external_ids, :namespace, :string, null: true
+    # The condition here allows two for two clients to share the same MCI ID
+    # https://github.com/greenriver/hmis-warehouse/pull/2933/files#r1164091887
+    add_index :external_ids,
+              [:source_type, :namespace, :value],
+              unique: true,
+              name: 'uidx_external_id_ns_value',
+              where: "(namespace <> 'ac_hmis_mci')"
```

Comment:
Should this also include namespace is not null?

Reply:
Should this also include `active`?  Maybe one gets set as inactive, and a new one added to cover the same namespace?

Reply:
> Should this also include active? Maybe one gets set as inactive, and a new one added to cover the same namespace?

Maybe? I guess I don't know how we would set the active value. I probably missed that spec

> Should this also include namespace is not null?

NULL values don't count towards a unique index afaik. So adding that to the condition shouldn't matter [(ref)](https://www.postgresql.org/docs/current/indexes-unique.html)

>  Null values are not considered equal. A multicolumn unique index will only reject cases where all indexed columns are equal in multiple rows.

Reply:
Yeah, there's no spec on `active` but you can set a remote credential as not active, so that we can turn them on and off if necessary.  It seems like there's a potential to want to replace a credential over time and mark previous ones as inactive.  Probably fine to leave as is for now.  

Null values not counting as equal seems to solve the second question.  Thanks!